### ISSUE TITLE
Add circular padding for convolution

### ISSUE THREAD

**Issue by HenryJia** [Theano/Theano#5639]
Circular convolution is needed as part of some models such as Neural Turing Machines. It would be nice to have it in as padding option for the convolution functions

**Comment by nouiz** [comment]
I added this to the more convolution GSoC project idea.

Do you link link to paper/code for that? It would help better define what you ask and they can be added in the code/doc when someone does it. Also, can you give more detail of the padding you are wanting? Can you confirm this is the same padding shape that we do normally, but that instead of padding with 0s, you want to pad with other values? Can you fill this information?

original filter

1 2 3
4 5 6
7 8 9

with padding =1, the padded filter should have shape of 5x5, but can you show the values you want?

**Comment by HenryJia** [comment]
Formula 8 of page 9 of https://arxiv.org/abs/1410.5401

For the example you stated I think it should be

9 7 8 9 7
3 1 2 3 1
6 4 5 6 4
9 7 8 9 7
3 1 2 3 1

Although the NTM paper only uses 1D circular conv

**Comment by nouiz** [comment]
Do you want that only for 1d yourself? If so, I think using indexing with join will be fast enough. I'm not so sure about 2d. So before doing more complicated solution, can you try the subtensor/join approch and benchmark it?

**Comment by HenryJia** [comment]
Yep, will do

**Comment by affanv14** [comment]
for 1d ,would a filter with shape 3x1 become 5x1?

**Comment by affanv14** [comment]
@HenryJia can you please pad this filter for padding=2 as well,I'll have a better idea overall then

**Comment by vikramnitin9** [comment]
Computing all indices modulo 3, it should be 

5 6 4 5 6 4 5
8 9 7 8 9 7 8
2 3 **1 2 3** 1 2
5 6 **4 5 6** 4 5
8 9 **7 8 9** 7 8
2 3 1 2 3 1 2
5 6 4 5 6 4 5

**Comment by hackintoshrao** [comment]
Would like to work on this, are you guys still open for PR on this?

**Comment by nouiz** [comment]
Yes, we are still open to PR on this. Keep us regularly updated on your
progress or if you stop working on this. We have others people working on
convolution and we need to all synchronize. Nobody else is currently
working on circular padding to my knowledge.

**Comment by hackintoshrao** [comment]
Thanks @nouiz , I'm on it and will keep updated.

**Comment by hackintoshrao** [comment]
What could be the best channel for micro discussions on the implementation?

**Comment by hackintoshrao** [comment]
Couldn't pick up the issue due to personal reasons, continuing from where I left off. 

I'm summarizing my understanding and next steps, 

- We are talking about reflection padding here. 
- Also you would like to first evaluate how degrading the performance is using the `subtensor/join approach` before going ahead with more complicated solution, I assume similar performance benchmark is also expected in https://github.com/Theano/Theano/issues/5827 . 

If the above mentioned understandings are right, I would be great if I can get some references to understand what exactly `subtensor/join approach` is, so that I could go ahead and benchmark them first.
